### ISSUE TITLE
only Train one epoch on CPU docker

### ISSUE THREAD

**Issue by zcy618** [facebookresearch/wav2letter#587]
hi, dear friends:
I am using CPU docker, I downloaded CPU docker from https://hub.docker.com/layers/wav2letter/wav2letter/cpu-latest/images/sha256-175ce0d0e052ed81efedf0227f722ea57d0eb024a83d916f12d73948c46a9512?context=explore

and I downloaded librispeech following the instruction at: https://github.com/facebookresearch/wav2letter/tree/master/tutorials/1-librispeech_clean

and I think I correctly configured the file train.cfg. now I start to run training, but I found there is only once poch as below:

root@b024bb2af669:~# ./wav2letter/build/Train train --flagsfile wav2letter/tutorials/1-librispeech_clean/train.cfg --logtostderr=1 --minloglevel=0
I0328 16:43:10.388114  4721 Train.cpp:59] Reading flags from file wav2letter/tutorials/1-librispeech_clean/train.cfg
I0328 16:43:10.389801  4721 Train.cpp:148] Gflags after parsing 
--flagfile=; --fromenv=; --tryfromenv=; --undefok=; --tab_completion_columns=80; --tab_completion_word=; --help=false; --helpfull=false; --helpmatch=; --helpon=; --helppackage=false; --helpshort=false; --helpxml=false; --version=false; --adambeta1=0.90000000000000002; --adambeta2=0.999; --am=; --am_decoder_tr_dropout=0; --am_decoder_tr_layerdrop=0; --am_decoder_tr_layers=1; --arch=network.arch; --archdir=/root/wav2letter/tutorials/1-librispeech_clean/; --attention=content; --attentionthreshold=2147483647; --attnWindow=no; --attnconvchannel=0; --attnconvkernel=0; --attndim=0; --batchsize=4; --beamsize=2500; --beamsizetoken=250000; --beamthreshold=25; --blobdata=false; --channels=1; --criterion=ctc; --critoptim=sgd; --datadir=/root/asr_dataset/; --dataorder=input; --decoderattnround=1; --decoderdropout=0; --decoderrnnlayer=1; --decodertype=wrd; --devwin=0; --emission_dir=; --emission_queue_size=3000; --enable_distributed=false; --encoderdim=0; --eosscore=0; --eostoken=false; --everstoredb=false; --fftcachesize=1; --filterbanks=40; --flagsfile=wav2letter/tutorials/1-librispeech_clean/train.cfg; --framesizems=25; --framestridems=10; --gamma=1; --gumbeltemperature=1; --input=flac; --inputbinsize=100; --inputfeeding=false; --isbeamdump=false; --iter=25; --itersave=false; --labelsmooth=0; --leftWindowSize=50; --lexicon=/root/asr_dataset/am/lexicon.txt; --linlr=-1; --linlrcrit=-1; --linseg=0; --lm=; --lm_memory=5000; --lm_vocab=; --lmtype=kenlm; --lmweight=0; --localnrmlleftctx=0; --localnrmlrightctx=0; --logadd=false; --lr=0.10000000000000001; --lr_decay=9223372036854775807; --lr_decay_step=9223372036854775807; --lrcosine=false; --lrcrit=0; --maxdecoderoutputlen=200; --maxgradnorm=1; --maxisz=9223372036854775807; --maxload=-1; --maxrate=10; --maxsil=50; --maxtsz=9223372036854775807; --maxword=-1; --melfloor=1; --memstepsize=10485760; --mfcc=false; --mfcccoeffs=13; --mfsc=true; --minisz=0; --minrate=3; --minsil=0; --mintsz=0; --momentum=0; --netoptim=sgd; --noresample=false; --nthread=4; --nthread_decoder=1; --nthread_decoder_am_forward=1; --numattnhead=8; --onorm=target; --optimepsilon=1e-08; --optimrho=0.90000000000000002; --outputbinsize=5; --pctteacherforcing=100; --pcttraineval=100; --pow=false; --pretrainWindow=0; --replabel=1; --reportiters=0; --rightWindowSize=50; --rndv_filepath=; --rundir=/root/asr_dataset/saved_models; --runname=librispeech_clean_trainlogs; --samplerate=16000; --sampletarget=0; --samplingstrategy=rand; --sclite=; --seed=0; --show=false; --showletters=false; --silscore=0; --smearing=none; --smoothingtemperature=1; --softwoffset=10; --softwrate=5; --softwstd=5; --sqnorm=true; --stepsize=9223372036854775807; --surround=|; --tag=; --target=tkn; --test=; --tokens=/root/asr_dataset/am/tokens.txt; --tokensdir=; --train=lists/train-clean-100.lst; --trainWithWindow=false; --transdiag=0; --unkscore=-inf; --use_memcache=false; --uselexicon=true; --usewordpiece=false; --valid=lists/dev-clean.lst; --weightdecay=0; --wordscore=0; --wordseparator=|; --world_rank=0; --world_size=1; --alsologtoemail=; --alsologtostderr=false; --colorlogtostderr=false; --drop_log_memory=true; --log_backtrace_at=; --log_dir=; --log_link=; --log_prefix=true; --logbuflevel=0; --logbufsecs=30; --logemaillevel=999; --logmailer=/bin/mail; --logtostderr=true; --max_log_size=1800; --minloglevel=0; --stderrthreshold=2; --stop_logging_if_full_disk=false; --symbolize_stacktrace=true; --v=0; --vmodule=; 
I0328 16:43:10.389942  4721 Train.cpp:149] Experiment path: /root/asr_dataset/saved_models/librispeech_clean_trainlogs
I0328 16:43:10.389957  4721 Train.cpp:150] Experiment runidx: 1
I0328 16:43:10.396713  4721 Train.cpp:194] Number of classes (network): 30
I0328 16:43:10.437093  4721 Train.cpp:201] Number of words: 34795
I0328 16:43:10.452620  4721 Train.cpp:215] Loading architecture file from /root/wav2letter/tutorials/1-librispeech_clean/network.arch
I0328 16:43:10.462795  4721 Train.cpp:247] [Network] Sequential [input -> (0) -> (1) -> (2) -> (3) -> (4) -> (5) -> (6) -> (7) -> (8) -> (9) -> (10) -> (11) -> (12) -> (13) -> (14) -> (15) -> (16) -> (17) -> (18) -> (19) -> (20) -> output]
	(0): View (-1 1 40 0)
	(1): Conv2D (40->256, 8x1, 2,1, SAME,SAME, 1, 1) (with bias)
	(2): ReLU
	(3): Conv2D (256->256, 8x1, 1,1, SAME,SAME, 1, 1) (with bias)
	(4): ReLU
	(5): Conv2D (256->256, 8x1, 1,1, SAME,SAME, 1, 1) (with bias)
	(6): ReLU
	(7): Conv2D (256->256, 8x1, 1,1, SAME,SAME, 1, 1) (with bias)
	(8): ReLU
	(9): Conv2D (256->256, 8x1, 1,1, SAME,SAME, 1, 1) (with bias)
	(10): ReLU
	(11): Conv2D (256->256, 8x1, 1,1, SAME,SAME, 1, 1) (with bias)
	(12): ReLU
	(13): Conv2D (256->256, 8x1, 1,1, SAME,SAME, 1, 1) (with bias)
	(14): ReLU
	(15): Conv2D (256->256, 8x1, 1,1, SAME,SAME, 1, 1) (with bias)
	(16): ReLU
	(17): Reorder (2,0,3,1)
	(18): Linear (256->512) (with bias)
	(19): ReLU
	(20): Linear (512->30) (with bias)
I0328 16:43:10.462837  4721 Train.cpp:248] [Network Params: 3900958]
I0328 16:43:10.462849  4721 Train.cpp:249] [Criterion] ConnectionistTemporalClassificationCriterion
I0328 16:43:10.462869  4721 Train.cpp:257] [Network Optimizer] SGD
I0328 16:43:10.462877  4721 Train.cpp:258] [Criterion Optimizer] SGD
I0328 16:43:10.914499  4721 W2lListFilesDataset.cpp:141] 28539 files found. 
I0328 16:43:10.915771  4721 Utils.cpp:102] Filtered 0/28539 samples
I0328 16:43:10.918277  4721 W2lListFilesDataset.cpp:62] Total batches (i.e. iters): 7135
I0328 16:43:10.974102  4721 W2lListFilesDataset.cpp:141] 2703 files found. 
I0328 16:43:10.974582  4721 Utils.cpp:102] Filtered 0/2703 samples
I0328 16:43:10.974965  4721 W2lListFilesDataset.cpp:62] Total batches (i.e. iters): 676
I0328 16:43:10.975749  4721 Train.cpp:544] Shuffling trainset
I0328 16:43:10.976809  4721 Train.cpp:551] Epoch 1 started!
I0328 16:44:39.312683  4721 Train.cpp:340] epoch:        1 | nupdates:           26 | lr: 2.600000 | lrcriterion: 0.000000 | runtime: 00:00:17 | bch(ms): 688.70 | smp(ms): 19.31 | fwd(ms): 195.19 | crit-fwd(ms): 11.97 | bwd(ms): 444.13 | optim(ms): 26.60 | loss:  347.57909 | train-TER: 99.92 | train-WER: 100.00 | lists/dev-clean.lst-loss:   55.02879 | lists/dev-clean.lst-TER: 100.00 | lists/dev-clean.lst-WER: 100.00 | avg-isz: 1301 | avg-tsz: 217 | max-tsz: 299 | hrs:    0.38 | thrpt(sec/sec): 75.60
I0328 16:44:39.546887  4721 Train.cpp:717] Finished training

could you help to tell what should I do please?
Thanks.

**Comment by zcy618** [comment]
could some experts to check this issue please? I do not find out why there is only one poch training.
Thanks

**Comment by YuanGongND** [comment]
I have exactly the same problem, but using a GPU docker.

**Comment by zcy618** [comment]
> I have exactly the same problem, but using a GPU docker.

Do you find the root cause?

**Comment by YuanGongND** [comment]
No, I am still finding the reason. I found that the training for the first epoch actually stops at the 26th iteration for an unknown reason (i.e., nupdates: 26).

**Comment by zcy618** [comment]
yes, me too.

So I am curious that how others run it?

**Comment by YuanGongND** [comment]
Oh. I think I found the problem. In train.cfg, line 25 `--iter=25`, it is 25 iterations, not 25 epochs. That's why the training stops at the 25th iteration (25t iterations are smaller than the iteration numbers of one epoch). Change it to a larger number solved my problem.

**Comment by tlikhomanenko** [comment]
@zcy618, @YuanGongND 

yep, we recently switched from epochs to updates in the `Train.cpp`, didn't have time to update tutorial yet. Right now `--iter` specifies number of updates, so just set this flag to some huge number. We are planning to reconsider this flag and maybe come back for it to be the number of epochs.
